glide-core/src/socket_listener.rs

```diff
@@ -924,23 +806,109 @@ pub fn start_socket_listener_internal<InitCallback>(
     init_callback: InitCallback,
     socket_path: Option<String>,
 ) where
-    InitCallback: FnOnce(Result<String, String>) + Send + 'static,
+    InitCallback: FnOnce(Result<String, String>) + Send + Clone + 'static,
 {
+    static INITIALIZED_SOCKETS: Lazy<Arc<RwLock<HashSet<String>>>> =
+        Lazy::new(|| Arc::new(RwLock::new(HashSet::new())));
+
+    let socket_path = socket_path.unwrap_or_else(get_socket_path);
+
+    {
+        // Optimize for already initialized
+        let initialized_sockets = INITIALIZED_SOCKETS
+            .read()
+            .expect("Failed to acquire sockets db read guard");
+        if initialized_sockets.contains(&socket_path) {
+            init_callback(Ok(socket_path.clone()));
+            return;
+        }
+    }
+
+    // Retry with write lock, will be dropped upon the function completion
+    let mut sockets_write_guard = INITIALIZED_SOCKETS
+        .write()
+        .expect("Failed to acquire sockets db write guard");
+    if sockets_write_guard.contains(&socket_path) {
+        init_callback(Ok(socket_path.clone()));
+        return;
+    }
+
+    let (tx, rx) = std::sync::mpsc::channel();
+    let socket_path_cloned = socket_path.clone();
+    let init_callback_cloned = init_callback.clone();
+    let tx_cloned = tx.clone();
     thread::Builder::new()
         .name("socket_listener_thread".to_string())
         .spawn(move || {
-            let runtime = Builder::new_current_thread().enable_all().build();
-            match runtime {
-                Ok(runtime) => {
-                    let mut listener = Disposable::new(SocketListener::new(
-                        socket_path.unwrap_or_else(get_socket_path),
-                    ));
-                    runtime.block_on(listener.listen_on_socket(init_callback));
+            let init_result = {
+                let runtime = Builder::new_current_thread().enable_all().build();
+                if let Err(err) = runtime {
+                    log_error(
+                        "listen_on_socket",
+                        format!("Error failed to create a new tokio thread: {err}"),
+                    );
+                    return Err(err);
                 }
-                Err(err) => init_callback(Err(err.to_string())),
+
+                runtime.unwrap().block_on(async move {
+                    let listener_socket = UnixListener::bind(socket_path_cloned.clone());
+                    if let Err(err) = listener_socket {
+                        log_error(
+                            "listen_on_socket",
+                            format!("Error failed to bind listening socket: {err}"),
+                        );
+                        return Err(err);
+                    }
+                    let listener_socket = listener_socket.unwrap();
+
+                    // signal initialization success
+                    init_callback(Ok(socket_path_cloned.clone()));
+                    let _ = tx.send(true);
```

Comment:
what do you mean by handle ? we could log maybe, but its so unimportant I would not bother
Otherwise, it is safer to process to the accept loop since it might serve other clients

Reply:
logging an error is one of way of handling it. Ignoring the return value of `send(..)` seems a bit weird to me

Reply:
the log will be redundant because it cannot happen - the rx will not be closed